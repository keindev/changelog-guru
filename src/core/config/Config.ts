export enum ServiceProvider {
    GitHub = 'github',
    GitLab = 'gitlab',
}

export enum ChangeLevel {
    Major = 'major',
    Minor = 'minor',
    Patch = 'patch',
}

export enum ExclusionType {
    AuthorLogin = 'authorLogin',
    CommitType = 'commitType',
    CommitScope = 'commitScope',
    CommitSubject = 'commitSubject',
}

export interface IConfigOptions {
    provider: ServiceProvider;
    filePath: string;
    types: Map<string, ChangeLevel>;
    exclusions: Map<ExclusionType, string[]>;
    plugins: Map<string, IPluginOption>;
}

export default class Config {
    public readonly filePath: string;
    public readonly provider: ServiceProvider;

    private types: Map<string, ChangeLevel>;
    private plugins: Map<string, IPluginOption>;
    private exclusions: Map<ExclusionType, string[]>;

    public constructor(options: IConfigOptions) {
        this.provider = options.provider;
        this.filePath = options.filePath;
        this.types = options.types;
        this.plugins = options.plugins;
        this.exclusions = options.exclusions;
    }

    public getPlugin(name: string): IPluginOption | undefined {
        return this.plugins.get(name);
    }

    public getPlugins(): [string, IPluginOption][] {
        return [...this.plugins.entries()];
    }

    public getTypes(): [string, ChangeLevel][] {
        return [...this.types.entries()];
    }

    public getExclusions(): [ExclusionType, string[]][] {
        return [...this.exclusions.entries()];
    }
}
