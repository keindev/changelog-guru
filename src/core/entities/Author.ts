import Entity, { Priority } from './Entity';

export default class Author extends Entity {
    readonly url: string;
    readonly avatar: string;

    #priority = 0;

    constructor(name: string, url: string, avatar: string) {
        super(`@${name}`);

        const avatarUrl = new URL(avatar);

        avatarUrl.searchParams.delete('s');
        avatarUrl.searchParams.append('s', `${40}`);
        this.url = url;
        this.avatar = `${url}`;
    }

    get priority(): Priority | number {
        return this.#priority;
    }

    contribute(value = 1): void {
        if (value > 0) this.#priority += value;
    }
}
