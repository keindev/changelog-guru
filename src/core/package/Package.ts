import writePkg from 'write-pkg';
import readPkg, { PackageJson } from 'read-pkg';
import { TaskTree } from 'tasktree-cli';
import * as semver from 'semver';

export enum DependencyRuleType {
    // https://docs.npmjs.com/files/package.json#engines
    Engines = 'engines',
    // https://docs.npmjs.com/files/package.json#dependencies
    Dependencies = 'dependencies',
    // https://docs.npmjs.com/files/package.json#devdependencies
    DevDependencies = 'devDependencies',
    // https://docs.npmjs.com/files/package.json#peerdependencies
    PeerDependencies = 'peerDependencies',
    // https://docs.npmjs.com/files/package.json#optionaldependencies
    OptionalDependencies = 'optionalDependencies',
}

export enum RestrictionRuleType {
    // https://docs.npmjs.com/files/package.json#bundleddependencies
    BundledDependencies = 'bundledDependencies',
    // https://docs.npmjs.com/files/package.json#os
    OS = 'os',
    // https://docs.npmjs.com/files/package.json#cpu
    CPU = 'cpu',
}

export type PackageDependenciesStory = [IPackageDependencies | undefined, IPackageDependencies | undefined];
export type PackageRestrictionsStory = [string[] | undefined, string[] | undefined];

export interface IPackageDependencies {
    [key: string]: string;
}

export default class Package {
    public static DEFAULT_VERSION = '0.0.1';
    public static DEFAULT_LICENSE = '';
    public static DEFAULT_REPOSITORY = '';

    private data: PackageJson;

    public constructor() {
        const task = TaskTree.add('Reading package.json');

        this.data = readPkg.sync({ normalize: false });

        if (!this.data.license) task.fail('Package license is not specified');
        if (!this.data.version) task.fail('Package version is not specified');
        if (!this.data.repository) task.fail('Package repository url is not specified');

        task.log(`Version: {bold ${this.getVersion()}}`);
        task.log(`Repository: {bold ${this.getRepository()}}`);
        task.complete('Package information:');
    }

    public getRepository(): string {
        const {
            data: { repository },
        } = this;

        if (typeof repository === 'string') return repository;
        if (typeof repository === 'object') return repository.url;

        return Package.DEFAULT_REPOSITORY;
    }

    public getVersion(): string {
        const { version } = this.data;

        return version
            ? semver.valid(semver.coerce(version) || Package.DEFAULT_VERSION) || Package.DEFAULT_VERSION
            : Package.DEFAULT_VERSION;
    }

    public getLicense(): string {
        return this.data.license || Package.DEFAULT_LICENSE;
    }

    public getDependenciesStory(type: DependencyRuleType, prevState: readPkg.PackageJson): PackageDependenciesStory {
        const { data } = this;
        let result: PackageDependenciesStory | undefined;

        switch (type) {
            case DependencyRuleType.Engines:
                result = [data.engines, prevState.engines];
                break;
            case DependencyRuleType.Dependencies:
                result = [data.dependencies, prevState.dependencies];
                break;
            case DependencyRuleType.DevDependencies:
                result = [data.devDependencies, prevState.devDependencies];
                break;
            case DependencyRuleType.PeerDependencies:
                result = [data.peerDependencies, prevState.peerDependencies];
                break;
            case DependencyRuleType.OptionalDependencies:
                result = [data.optionalDependencies, prevState.optionalDependencies];
                break;
            default:
                TaskTree.fail(`Unexpected dependency group type: {bold ${type}}`);
                break;
        }

        return result as PackageDependenciesStory;
    }

    public getRestrictionsStory(type: RestrictionRuleType, prevState: readPkg.PackageJson): PackageRestrictionsStory {
        const { data } = this;
        let result: PackageRestrictionsStory | undefined;

        switch (type) {
            case RestrictionRuleType.BundledDependencies:
                result = [data.bundledDependencies, prevState.bundledDependencies];
                break;
            case RestrictionRuleType.CPU:
                result = [data.cpu, prevState.cpu];
                break;
            case RestrictionRuleType.OS:
                result = [data.os, prevState.os];
                break;
            default:
                TaskTree.fail(`Unexpected restriction group type: {bold ${type}}`);
                break;
        }

        return result as PackageRestrictionsStory;
    }

    public async incrementVersion(major: number, minor: number, patch: number): Promise<void> {
        const task = TaskTree.add(`Updating package version`);
        const current = this.getVersion();
        let next: string | undefined;

        if (major) next = semver.inc(current, 'major') || undefined;
        if (!major && minor) next = semver.inc(current, 'minor') || undefined;
        if (!major && !minor && patch) next = semver.inc(current, 'patch') || undefined;
        if (next) {
            this.data.version = next;
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            await writePkg(this.data as any);
            task.complete(`Package version updated to {bold ${next}}`);
        } else {
            task.fail(`New package version {bold.underline ${next}} is invalid or less (see https://semver.org/)`);
        }
    }
}
