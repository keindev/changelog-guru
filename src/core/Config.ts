import { cosmiconfig } from 'cosmiconfig';
import deepmerge from 'deepmerge';
import path from 'path';
import TaskTree from 'tasktree-cli';
import { fileURLToPath } from 'url';

import { ChangeLevel } from './entities/Entity';
import { BaseRule, IRule, IRuleConfig, Rule } from './rules/BaseRule';
import HighlightRule from './rules/HighlightRule';
import MarkRule from './rules/MarkRule';
import PackageStatisticRenderRule from './rules/PackageStatisticRenderRule';
import ScopeRenameRule from './rules/ScopeRenameRule';
import SectionGroupRule from './rules/SectionGroupRule';

export enum Exclusion {
  AuthorLogin = 'authorLogin',
  CommitType = 'commitType',
  CommitScope = 'commitScope',
  CommitSubject = 'commitSubject',
}

export enum GitServiceProvider {
  GitHub = 'github',
  GitLab = 'gitlab',
}

export interface IConfigOptions {
  branch?: string;
  bump?: boolean;
  output?: string;
  provider?: GitServiceProvider;
}

export interface IChangelogConfig {
  branch: string;
  changes: {
    [key in ChangeLevel]: string[];
  };
  output: {
    exclude: { [key in Exclusion]: string[] };
    filePath: string;
  };
  provider: GitServiceProvider;
  rules: {
    [key in Rule]: IRuleConfig;
  };
}

export class Config {
  #branch = '';
  #exclusions: [Exclusion, string[]][] = [];
  #filePath = 'CHANGELOG.md';
  #isInitialized = false;
  #options: IConfigOptions;
  #provider = GitServiceProvider.GitHub;
  #rules: IRule[] = [];
  #types: [string, ChangeLevel][] = [];

  constructor(options?: IConfigOptions) {
    if (options?.provider && !Object.values(GitServiceProvider).includes(options.provider)) {
      TaskTree.fail('Service provider not supported');
    }

    this.#options = options ?? {};
  }

  get provider(): GitServiceProvider {
    return this.#provider;
  }

  get branch(): string {
    return this.#branch;
  }

  get filePath(): string {
    return this.#filePath;
  }

  get exclusions(): [Exclusion, string[]][] {
    return this.#exclusions;
  }

  get types(): [string, ChangeLevel][] {
    return this.#types;
  }

  get rules(): IRule[] {
    return this.#rules;
  }

  get bump(): boolean {
    return !!this.#options.bump;
  }

  async init(): Promise<void> {
    if (!this.#isInitialized) {
      const dirname = path.dirname(fileURLToPath(import.meta.url));
      const task = TaskTree.add('Reading configuration file...');
      const explorer = cosmiconfig('changelog-guru');
      const baseConf = await explorer.load(path.join(dirname, '../../.changelogrc.default.yml'));
      const userConf = await explorer.search();

      if (baseConf?.config && !baseConf.isEmpty) {
        const config = deepmerge<IChangelogConfig>(baseConf.config, userConf?.config ?? {});
        const filePath = path.relative(process.cwd(), userConf?.filepath ?? baseConf.filepath);

        this.#types = this.getTypes(config.changes);
        this.#rules = this.getRules(config.rules);
        this.#provider = this.#options?.provider ?? config.provider;
        this.#branch = this.#options?.branch ?? config.branch;
        this.#filePath = this.#options?.output ?? config.output.filePath;
        this.#exclusions = Object.entries(config.output.exclude ?? {}).map(([name, rules]) => {
          if (!Object.values(Exclusion).includes(name as Exclusion)) {
            task.fail(`Unexpected exclusion name: {bold ${name}}`);
          }

          return [name as Exclusion, [...new Set(rules)]];
        });

        task.complete(`Configuration initialized: {bold ${filePath}}`);
      } else {
        task.fail('Default configuration file not found');
      }
    } else {
      this.#isInitialized = true;
    }
  }

  private getRules(configs: IChangelogConfig['rules']): IRule[] {
    const map = {
      [Rule.Highlight]: HighlightRule,
      [Rule.Mark]: MarkRule,
      [Rule.PackageStatisticRender]: PackageStatisticRenderRule,
      [Rule.ScopeRename]: ScopeRenameRule,
      [Rule.SectionGroup]: SectionGroupRule,
    };

    return [...Object.values(Rule)].map(rule => new (map[rule] as typeof BaseRule)(configs[rule]) as IRule);
  }

  private getTypes(changes: IChangelogConfig['changes']): [string, ChangeLevel][] {
    const levels = Object.values(ChangeLevel);

    return Object.entries(changes).reduce((acc, [level, names]) => {
      if (!Array.isArray(names)) TaskTree.fail(`Names of change level "${level}" must be array`);
      if (!levels.includes(level as ChangeLevel)) TaskTree.fail(`Unexpected level "${level}" of changes`);

      names.forEach(name => acc.push([name, level as ChangeLevel]));

      return acc;
    }, [] as [string, ChangeLevel][]);
  }
}
