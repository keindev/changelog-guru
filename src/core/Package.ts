import readPkg, { PackageJson } from 'read-pkg';
import { coerce, compare, inc, SemVer, valid } from 'semver';
import { TaskTree } from 'tasktree-cli';
import writePkg from 'write-pkg';

import { Compare } from './entities/Entity';

export enum Dependency {
  // https://docs.npmjs.com/files/package.json#engines
  Engines = 'engines',
  // https://docs.npmjs.com/files/package.json#dependencies
  Dependencies = 'dependencies',
  // https://docs.npmjs.com/files/package.json#devdependencies
  DevDependencies = 'devDependencies',
  // https://docs.npmjs.com/files/package.json#peerdependencies
  PeerDependencies = 'peerDependencies',
  // https://docs.npmjs.com/files/package.json#optionaldependencies
  OptionalDependencies = 'optionalDependencies',
}

export enum Restriction {
  // https://docs.npmjs.com/files/package.json#bundleddependencies
  BundledDependencies = 'bundledDependencies',
  // https://docs.npmjs.com/files/package.json#os
  OS = 'os',
  // https://docs.npmjs.com/files/package.json#cpu
  CPU = 'cpu',
}

export enum DependencyChangeType {
  Added = 'added',
  Bumped = 'bumped',
  Changed = 'changed',
  Downgraded = 'downgraded',
  Removed = 'removed',
  Unchanged = 'unchanged',
}

const VERSION_CHANGES_MAP = {
  [Compare.Less]: DependencyChangeType.Downgraded,
  [Compare.More]: DependencyChangeType.Bumped,
  [Compare.Equal]: DependencyChangeType.Unchanged,
};
const RESTRICTION_CHANGES_MAP = {
  [Compare.Less]: DependencyChangeType.Changed,
  [Compare.More]: DependencyChangeType.Changed,
  [Compare.Equal]: DependencyChangeType.Unchanged,
};

export interface IPackageChange {
  name: string;
  type: DependencyChangeType;
  value?: string;
  link?: string;
  version?: SemVer;
  prevValue?: string;
  prevVersion?: SemVer;
}

export default class Package {
  #data: PackageJson;

  constructor() {
    const task = TaskTree.add('Reading package.json');

    this.#data = readPkg.sync({ normalize: false });

    if (!this.#data.license) task.fail('Package license is not specified');
    if (!this.#data.version) task.fail('Package version is not specified');
    if (!this.#data.repository) task.fail('Package repository url is not specified');

    task.log(`Version: {bold ${this.version}}`);
    task.log(`Repository: {bold ${this.repository}}`);
    task.complete('Package information:');
  }

  get repository(): string {
    const { repository } = this.#data;

    if (!repository) TaskTree.fail('Package repository is empty');

    return typeof repository === 'object' ? repository.url : repository;
  }

  get version(): string {
    const semanticVersion = coerce(this.#data.version) ?? '';
    const validVersion = valid(semanticVersion);

    if (!semanticVersion) TaskTree.fail('Package version is empty or not SemVer (see https://semver.org/)');
    if (!validVersion) TaskTree.fail('Package version is not a valid SemVer (see https://semver.org/)');

    return validVersion;
  }

  get license(): string {
    return this.#data.license ?? '';
  }

  async bump(major: number, minor: number, patch: number): Promise<void> {
    const task = TaskTree.add('Updating package version');
    let version: string | null | undefined;

    if (major) version = inc(this.version, 'major');
    if (!major && minor) version = inc(this.version, 'minor');
    if (!major && !minor && patch) version = inc(this.version, 'patch');

    if (version) {
      this.#data.version = version;

      await writePkg({ ...(this.#data as { [key: string]: string }) });
      task.complete(`Package version updated to {bold ${version}}`);
    } else {
      task.fail('New package version is invalid or less (see https://semver.org/)');
    }
  }

  getChanges(property: Dependency | Restriction, prevValues: { [key: string]: string } | string[]): IPackageChange[] {
    return Array.isArray(prevValues)
      ? this.getRestrictionsChanges(property as Restriction, prevValues)
      : this.getDependenciesChanges(property as Dependency, prevValues);
  }

  private getDependenciesChanges(property: Dependency, dependencies: { [key: string]: string }): IPackageChange[] {
    const changes: IPackageChange[] = [];
    const currDeps = this.#data[property];
    const prevDeps = new Map(Object.entries(dependencies));

    if (currDeps) {
      const getLink = (type: Dependency, name: string, ver?: string): string | undefined => {
        const link = `https://www.npmjs.com/package/${name}/v/${ver}`;

        return type !== Dependency.Engines && ver ? link : undefined;
      };

      Object.entries(currDeps).forEach(([name, value]) => {
        const version = coerce(value);

        if (version) {
          const prevValue = prevDeps.get(name);
          const prevVersion = coerce(prevValue) ?? undefined;
          const link = getLink(property, name, version.version);
          let type = prevVersion ? VERSION_CHANGES_MAP[compare(version, prevVersion)] : DependencyChangeType.Added;

          if (type === DependencyChangeType.Unchanged && value !== prevValue) {
            type = DependencyChangeType.Changed;
          }

          changes.push({ name, value, type, version, link, prevValue, prevVersion });
          prevDeps.delete(name);
        }
      });

      prevDeps.forEach((prevValue, name) => {
        const prevVersion = coerce(prevValue) ?? undefined;
        const link = getLink(property, name, prevVersion?.version);

        changes.push({ name, type: DependencyChangeType.Removed, link, prevValue, prevVersion });
      });
    }

    return [...changes.values()].filter(change => change.type !== DependencyChangeType.Unchanged);
  }

  private getRestrictionsChanges(property: Restriction, prevRestrictions: string[]): IPackageChange[] {
    const changes: IPackageChange[] = [];
    const currRestrictions: string[] = this.#data[property] ?? [];
    const restrictions = [...prevRestrictions];
    const getName = (name: string): string => (name[0] === '!' ? name.slice(1) : name);

    changes.push(
      ...currRestrictions.map(value => {
        const index = restrictions.indexOf(value);
        const prevValue = restrictions[index];
        const type = prevValue
          ? RESTRICTION_CHANGES_MAP[value.localeCompare(prevValue) as Compare]
          : DependencyChangeType.Added;

        if (prevValue) restrictions.splice(index, 1);

        return { name: getName(value), value, type, prevValue };
      }),
      ...restrictions.map(prevValue => ({ name: getName(prevValue), type: DependencyChangeType.Removed, prevValue }))
    );

    return [...changes.values()].filter(change => change.type !== DependencyChangeType.Unchanged);
  }
}
