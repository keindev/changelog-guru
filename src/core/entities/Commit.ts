import LookupManager from 'string-lookup-manager';

import { splitHeadline } from '../../utils/commit.js';
import { wrap } from '../../utils/markdown.js';
import { IAuthor } from './Author.js';
import Entity, { ChangeLevel, Compare, IEntity, Priority } from './Entity.js';

export enum CommitChangeType {
  BreakingChanges = 1,
  Deprecated = 2,
  Important = 4,
  Default = 8,
}

const COMMIT_SHOT_NAME_LENGTH = 7;

export interface ICommit extends IEntity {
  readonly accents: string[];
  readonly author: IAuthor;
  readonly body: readonly string[];
  changeType: CommitChangeType;
  readonly scope?: string;
  readonly shortName: string;
  subject: string;
  readonly timestamp: number;
  readonly type: string;

  readonly url: string;

  accent(text: string): void;
  is(status: CommitChangeType): boolean;
  replacement(value: string, position: number): void;
}

export interface ICommitOptions {
  author: IAuthor;
  body?: string;
  hash: string;
  headline: string;
  timestamp: number;
  url: string;
}

export default class Commit extends Entity implements ICommit {
  readonly author: IAuthor;
  readonly body: readonly string[] = [];
  readonly scope?: string;
  readonly timestamp: number;
  readonly url: string;

  #accents = new Set<string>();
  #changeType = CommitChangeType.Default;
  #replacements = new LookupManager();
  #subject = '';
  #type?: string;

  constructor({ hash, timestamp, headline, body, url, author }: ICommitOptions) {
    super(hash);

    this.timestamp = timestamp;
    this.url = url;
    this.author = author;

    const [type, scope, subject] = splitHeadline(headline);

    if (type) this.#type = type.toLocaleLowerCase();
    if (scope) this.scope = scope;
    if (subject) this.#subject = subject;
    if (body) this.body = body.split('\n').map(line => line.trim());
  }

  static compare(a: ICommit, b: ICommit): Compare {
    const { scope: x } = a;
    const { scope: y } = b;
    let result = super.compare(a, b);

    if (x && !y) result--;
    if (!x && y) result++;
    if (x && y) result = x.localeCompare(y);
    if (result === Compare.Equal) result = a.timestamp - b.timestamp;

    return Math.min(Math.max(result, Compare.Less), Compare.More);
  }

  get accents(): string[] {
    return [...this.#accents.values()];
  }

  get type(): string {
    return this.#type ?? '';
  }

  get priority(): Priority {
    let priority = super.priority;

    if (this.is(CommitChangeType.BreakingChanges)) priority += Priority.High;
    if (this.is(CommitChangeType.Deprecated)) priority += Priority.Medium;
    if (this.is(CommitChangeType.Important)) priority += Priority.Low;

    return priority;
  }

  get subject(): string {
    return this.#replacements.replace(this.#subject, item => wrap(item.value));
  }

  get changeType(): CommitChangeType {
    return this.#changeType;
  }

  set changeType(type: CommitChangeType) {
    this.#changeType = this.#changeType | type;

    if (this.is(CommitChangeType.Deprecated)) this.level = ChangeLevel.Minor;
    if (this.is(CommitChangeType.BreakingChanges)) this.level = ChangeLevel.Major;
  }

  get shortName(): string {
    return this.name.substr(0, COMMIT_SHOT_NAME_LENGTH);
  }

  accent(text: string): void {
    this.#accents.add(text);
  }

  is(type: CommitChangeType): boolean {
    return !!(this.#changeType & type);
  }

  replacement(value: string, position: number): void {
    this.#replacements.add(value, position);
  }
}
