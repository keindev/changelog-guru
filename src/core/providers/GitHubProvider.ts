import Provider from 'gh-gql';
import { ICommit as IGitHubCommit } from 'gh-gql/lib/queries/Commit';
import { TaskTree } from 'tasktree-cli';
import { PackageJson } from 'type-fest';

import { GitServiceProvider } from '../Config';
import Author, { IAuthor } from '../entities/Author';
import Commit, { ICommit } from '../entities/Commit';
import GitProvider, { IGitProviderOptions } from './GitProvider';

export default class GitHubProvider extends GitProvider {
  #provider: Provider;
  #authors = new Map<number, Author>();

  constructor(url: string, branch: IGitProviderOptions['branch']) {
    super({ type: GitServiceProvider.GitHub, url, branch });

    this.#provider = new Provider();
  }

  async getLastChangeDate(dev?: boolean): Promise<Date> {
    const lastCommit = await this.#provider.query.commit.getLastCommit({
      owner: this.owner,
      repository: this.repository,
      branch: dev ? this.branch.dev : this.branch.main,
    });

    return new Date(lastCommit?.committedDate ?? Date.now());
  }

  async getCommits(since: Date): Promise<ICommit[]> {
    const commits: ICommit[] = [];
    const list = await this.#provider.query.commit.getList({
      owner: this.owner,
      repository: this.repository,
      branch: this.branch.dev,
      since: since.toISOString(),
    });

    commits.push(...list.map(commit => this.parseCommit(commit)));

    return commits;
  }

  async getPreviousPackage(since: Date): Promise<PackageJson> {
    const data = await this.getPackage(this.branch.main, since);

    return data;
  }

  async getCurrentPackage(since: Date): Promise<PackageJson> {
    const data = await this.getPackage(this.branch.dev, since);

    return data;
  }

  private async getPackage(branch: string, since: Date): Promise<PackageJson> {
    const { owner, repository, package: filePath } = this;
    const task = TaskTree.add(`Loading {bold package.json} from {bold ${branch}}...`);
    let data: PackageJson = {};

    const id = await this.#provider.query.file.getId({
      until: since.toISOString(),
      branch,
      owner,
      repository,
      filePath,
    });

    if (id) {
      const text = await this.#provider.query.file.getContent({ filePath, owner, repository, oid: id });

      if (text) {
        data = JSON.parse(text);
        task.complete(`File {bold package.json} is loaded for {bold ${branch}}`);
      }
    } else {
      task.skip(`Branch {bold ${branch}} did not contain {bold package.json}`);
    }

    return data;
  }

  private parseCommit(commit: IGitHubCommit): ICommit {
    const { url, oid: hash, messageHeadline: headline, messageBody: body } = commit;
    const timestamp = new Date(commit.committedDate).getTime();
    const author = this.parseAuthor(commit.author);

    return new Commit({ hash, headline, body, author, timestamp, url });
  }

  private parseAuthor({ avatarUrl, user: { databaseId, login, url } }: IGitHubCommit['author']): IAuthor {
    const author = this.#authors.get(databaseId) ?? new Author({ login, url, avatar: avatarUrl });

    if (!this.#authors.has(databaseId)) this.#authors.set(databaseId, author);

    return author;
  }
}
