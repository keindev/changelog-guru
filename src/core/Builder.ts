import { promises as fs } from 'fs';
import Package from 'package-json-helper';
import { PackageDependency, PackageRestriction } from 'package-json-helper/lib/types';
import path from 'path';
import TaskTree from 'tasktree-cli';

import md from '../utils/markdown.js';
import { findSame } from '../utils/text.js';
import { Config, GitServiceProvider } from './Config.js';
import Commit, { ICommit } from './entities/Commit.js';
import Message from './entities/Message.js';
import Section, { ISection } from './entities/Section.js';
import GitHubProvider from './providers/GitHubProvider.js';
import GitLabProvider from './providers/GitLabProvider.js';
import { IGitProvider } from './providers/GitProvider.js';
import State from './State.js';

export default class Builder {
  #config: Config;
  #package: Package;
  #state?: State;

  constructor(config: Config) {
    this.#config = config;
    this.#package = new Package();
  }

  async build(): Promise<void> {
    await this.#config.init();
    await this.#package.read();

    if (this.#package.repository?.url) {
      const provider = new {
        [GitServiceProvider.GitHub]: GitHubProvider,
        [GitServiceProvider.GitLab]: GitLabProvider,
      }[this.#config.provider](this.#package.repository.url, this.#config.branch);

      await this.read(provider);
      await this.write(provider);
    } else {
      TaskTree.fail('Package repository url is not defined!');
    }
  }

  private async bumpPackage(provider: IGitProvider): Promise<void> {
    if (this.#config.bump && this.#package.version && this.#state) {
      const task = TaskTree.add('Updating package version');
      const date = await provider.getLastChangeDate(true);
      const { version: prevVersion } = await provider.getCurrentPackage(date);
      const { version } = this.#package;

      if (prevVersion && prevVersion !== version) {
        task.skip(
          `Package version is already changed from {bold ${prevVersion}(${provider.branch.dev})} to {bold ${version}}`
        );
      } else {
        const [major, minor, patch] = this.#state.changesLevels;

        if (major || minor || patch) {
          this.#package.bump({ major, minor, patch });
          await this.#package.save();
          task.complete(`Package version updated to {bold ${this.#package.version}}`);
        } else {
          task.skip('Package version does not change');
        }
      }

      task.clear();
    }
  }

  private async read(provider: IGitProvider): Promise<void> {
    const stage = TaskTree.add('Loading repository changes...');

    try {
      const date = await provider.getLastChangeDate();
      const commits = await provider.getCommits(date);
      const previousPackage = await provider.getPreviousPackage(date);

      if (this.#package.license) {
        const state = new State(this.#package.license, previousPackage.license);

        [...Object.values(PackageDependency), ...Object.values(PackageRestriction)].forEach(name => {
          state.addChanges(name, this.#package.getChanges(name, previousPackage));
        });

        commits.forEach(commit => state.addCommit(commit));
        this.#state = state;

        stage.clear();
        stage.log(`Branch {bold ${this.#config.branch}} last changes at: {bold ${date.toLocaleString()}}`);
        stage.log(`{bold ${commits.length}} commits loaded`);
        stage.complete('Release information:');
      } else {
        stage.fail('Package license is not defined');
      }
    } catch (error) {
      stage.error(error, true);
    }
  }

  private renderCommits(commits: ICommit[]): string[] {
    const groups = new Map<string, { accents: Set<string>; links: string[]; subject: string }>();

    commits.forEach(commit => {
      const subject = findSame(commit.subject, [...groups.keys()]) ?? commit.subject;
      const { accents, links } = groups.get(subject) ?? { subject, accents: new Set(), links: [] };

      commit.accents.forEach(accent => accents.add(accent));
      links.push(md.commit(commit.shortName, commit.url));
      groups.set(commit.subject, { subject, accents, links });
    });

    return [
      ...[...groups.values()].map(({ subject, accents, links }) =>
        md.list(
          [
            accents.size ? md.strong(`[${[...accents.values()].map(md.capitalize).join(', ')}]`) : '',
            md.capitalize(subject),
            ...links,
          ]
            .filter(Boolean)
            .join(' ')
        )
      ),
      '',
    ];
  }

  private renderSection(section: ISection, level = 1): string {
    const sections = section.sections.filter(Section.filter);
    const commits = section.commits.filter(Commit.filter);
    const messages = section.messages.filter(Message.filter);
    const output = section.isDetails ? [md.summary(section.title)] : [md.title(section.title, level)];

    if (messages.length) output.push(...messages.map(message => message.text), '');
    if (sections.length) output.push(...sections.map(subsection => this.renderSection(subsection, level + 1)));
    if (sections.length && commits.length) output.push(md.title('Others', level + 1));
    if (commits.length) output.push(...this.renderCommits(commits));

    return section.isDetails ? md.details(output.join('\n')) : output.join('\n');
  }

  private async write(provider: IGitProvider): Promise<void> {
    if (this.#state) {
      const task = TaskTree.add('Generate changelog...');

      this.#state.updateCommitsChangeLevel(this.#config.types);
      this.#state.ignore(this.#config.exclusions);
      await this.#state.modify(this.#config.rules);

      const data = this.#state.sections.map(subsection => this.renderSection(subsection));
      const filePath = path.resolve(process.cwd(), this.#config.filePath);

      if (this.#state.authors) {
        data.push(md.contributors(this.#state.authors.map(({ name, avatar, url }) => md.image(name, avatar, url))), '');
      }

      await fs.writeFile(filePath, data.join('\n'));
      await this.bumpPackage(provider);

      task.complete('Changelog generated!');
    }
  }
}
