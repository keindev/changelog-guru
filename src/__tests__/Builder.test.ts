// @see https://github.com/facebook/jest/issues/9430
// eslint-disable-next-line node/no-extraneous-import
import { jest } from '@jest/globals';
import fs from 'fs';
import Package from 'package-json-helper';
import path from 'path';

import Builder from '../core/Builder';
import { Config } from '../core/Config';
import Author from '../core/entities/Author';
import Commit from '../core/entities/Commit';
import GitHubProvider from '../core/providers/GitHubProvider';

jest.useFakeTimers();
jest.mock('../core/providers/GitHubProvider.ts');

describe('Builder', () => {
  const hash = new Array(40).fill(0).join('').slice(0, -1);
  const url = `https://github.com/keindev/changelog-guru/commit/${hash}`;
  const headlines = [
    'test(Builder): test commit',
    'feat: feat commit',
    'feat: feat commit',
    'feat: another feat commit',
    'perf(core): pref subject',
    'perf(core): pref subject',
    'build: ignored build commit',
  ];
  const author = new Author({
    login: 'keindev',
    url: 'https://github.com/keindev',
    avatar: 'https://avatars.githubusercontent.com/u/4527292',
  });
  const commits = headlines.map(
    (headline, index) =>
      new Commit({
        author,
        url: `${url}${index}`,
        hash: `${hash}${index}`,
        headline,
        timestamp: new Date(new Date(0).getTime() + index * 1000).getTime(),
      })
  );

  beforeAll(() => {
    const data = JSON.parse(fs.readFileSync(path.resolve(process.cwd(), 'package.json'), 'utf8'));

    jest.spyOn(GitHubProvider.prototype, 'getLastChangeDate').mockImplementation(() => Promise.resolve(new Date(0)));
    jest.spyOn(GitHubProvider.prototype, 'getCommits').mockImplementation(() => Promise.resolve([...commits]));
    jest
      .spyOn(GitHubProvider.prototype, 'getPreviousPackage')
      .mockImplementation(() => Promise.resolve(new Package(data)));
    jest
      .spyOn(GitHubProvider.prototype, 'getCurrentPackage')
      .mockImplementation(() => Promise.resolve(new Package(data)));
    jest.spyOn(fs.promises, 'readFile').mockImplementation(filePath => {
      const basename = path.basename(filePath as string);
      let content = '';

      if (basename === 'package.json') content = JSON.stringify(data);

      return Promise.resolve(content);
    });
  });

  it('build', async () => {
    const config = new Config() as jest.Mocked<Config>;
    const builder = new Builder(config);
    const files: [string, string][] = [];

    jest.spyOn(fs.promises, 'writeFile').mockImplementation((name, data) => {
      files.push([path.relative(process.cwd(), name.toString()), data.toString()]);

      return Promise.resolve();
    });

    await builder.build();

    expect(files).toMatchSnapshot();
  });
});
