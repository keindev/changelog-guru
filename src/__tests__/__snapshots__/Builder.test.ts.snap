// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder build 1`] = `
"# Features

- Feat commit [\`0000000\`](https://github.com/keindev/changelog-guru/commit/0000000000000000000000000000000000000001) [\`0000000\`](https://github.com/keindev/changelog-guru/commit/0000000000000000000000000000000000000002)
- Another feat commit [\`0000000\`](https://github.com/keindev/changelog-guru/commit/0000000000000000000000000000000000000003)

# Internal changes

- **[Builder]** Test commit [\`0000000\`](https://github.com/keindev/changelog-guru/commit/0000000000000000000000000000000000000000)

# Performance Improvements

- **[Core]** Pref subject [\`0000000\`](https://github.com/keindev/changelog-guru/commit/0000000000000000000000000000000000000004) [\`0000000\`](https://github.com/keindev/changelog-guru/commit/0000000000000000000000000000000000000005)

---

# Contributors

[![@keindev](https://avatars.githubusercontent.com/u/4527292?s=40)](https://github.com/keindev)
"
`;
